log.go

```diff
@@ -2,12 +2,119 @@ package didplc
 
 import (
 	"fmt"
+	"sync"
 	"time"
 
-	"github.com/bluesky-social/indigo/atproto/crypto"
 	"github.com/bluesky-social/indigo/atproto/syntax"
 )
 
+type OpStatus struct {
+	CreatedAt   time.Time // the only immutable field here
+	Nullified   bool
+	LastChild   string // CID
+	AllowedKeys []string
+}
+
+// Note: LogValidationContext is designed such that it could later be turned into an interface,
+// optionally backed by a db rather than in-memory
+// Note: ops are globally unique by CID, so opStatus map can be shared across all DIDs
+type LogValidationContext struct {
+	head     map[string]string    // DID -> CID, tracks most recent valid op for a particular DID
+	opStatus map[string]*OpStatus // CID -> OpStatus
+	lock     sync.RWMutex
+}
+
+func (c *LogValidationContext) GetValidationContext(did string, cid string) (string, *OpStatus, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	head, exists := c.head[did]
+	if !exists {
+		if cid != "" {
+			return "", nil, fmt.Errorf("DID not found")
+		}
+		return head, nil, nil // Not an error condition! just means DID is not created yet
+	}
+	status := c.opStatus[cid]
+	if status == nil {
+		return "", nil, fmt.Errorf("CID not found")
+	}
+
+	// make a deep copy of the status struct so that concurrent mutations are safe
+	statusCopy := *status
+	statusCopy.AllowedKeys = make([]string, len(status.AllowedKeys))
+	copy(statusCopy.AllowedKeys, status.AllowedKeys)
+
+	return head, &statusCopy, nil
+}
+
+func (c *LogValidationContext) CommitValidOperation(did string, head string, prevStatus *OpStatus, op Operation, createdAt time.Time, keyIndex int) error {
```

Comment:
could use a doc comment: this method appends a new operation for the indicated DID. the operation is expected to have been validated (eg, signatures and required fields checked), but this method will validate consistency with the current oplog state for the DID (eg, 72hr revision window and nullification state)